File: packages/analyzer/src/core/importParser.ts
import {
  readStringLiteral,
  skipWhitespaceAndComments,
  startsWithToken
} from "./text.js";

interface ParseResult {
  nextIndex: number;
  specifier: string | null;
}

function scanToBoundary(source: string, start: number): number {
  let cursor = start;
  let depth = 0;

  while (cursor < source.length) {
    cursor = skipWhitespaceAndComments(source, cursor);
    if (cursor >= source.length) {
      return cursor;
    }

    const char = source[cursor];
    if (char === "\"" || char === "'" || char === "`") {
      cursor = readStringLiteral(source, cursor).end;
      continue;
    }

    if (char === "(" || char === "{" || char === "[") {
      depth += 1;
      cursor += 1;
      continue;
    }

    if (char === ")" || char === "}" || char === "]") {
      depth = Math.max(0, depth - 1);
      cursor += 1;
      continue;
    }

    if ((char === ";" || char === "\n") && depth === 0) {
      return cursor + 1;
    }

    cursor += 1;
  }

  return source.length;
}

function parseImport(source: string, start: number): ParseResult {
  let cursor = skipWhitespaceAndComments(source, start + "import".length);

  if (source[cursor] === "(") {
    cursor = skipWhitespaceAndComments(source, cursor + 1);
    if (source[cursor] === "\"" || source[cursor] === "'") {
      const literal = readStringLiteral(source, cursor);
      const nextIndex = scanToBoundary(source, literal.end);
      return { nextIndex, specifier: literal.value };
    }

    return { nextIndex: scanToBoundary(source, cursor), specifier: null };
  }

  if (source[cursor] === "\"" || source[cursor] === "'") {
    const literal = readStringLiteral(source, cursor);
    return { nextIndex: scanToBoundary(source, literal.end), specifier: literal.value };
  }

  while (cursor < source.length) {
    cursor = skipWhitespaceAndComments(source, cursor);
    if (cursor >= source.length) {
      break;
    }

    if (startsWithToken(source, cursor, "from")) {
      cursor = skipWhitespaceAndComments(source, cursor + "from".length);
      if (source[cursor] === "\"" || source[cursor] === "'") {
        const literal = readStringLiteral(source, cursor);
        return { nextIndex: scanToBoundary(source, literal.end), specifier: literal.value };
      }

      return { nextIndex: scanToBoundary(source, cursor), specifier: null };
    }

    const char = source[cursor];
    if (char === "\"" || char === "'" || char === "`") {
      cursor = readStringLiteral(source, cursor).end;
      continue;
    }

    if (char === ";" || char === "\n") {
      return { nextIndex: cursor + 1, specifier: null };
    }

    cursor += 1;
  }

  return { nextIndex: cursor, specifier: null };
}

function parseExport(source: string, start: number): ParseResult {
  let cursor = skipWhitespaceAndComments(source, start + "export".length);

  while (cursor < source.length) {
    cursor = skipWhitespaceAndComments(source, cursor);
    if (cursor >= source.length) {
      break;
    }

    if (startsWithToken(source, cursor, "from")) {
      cursor = skipWhitespaceAndComments(source, cursor + "from".length);
      if (source[cursor] === "\"" || source[cursor] === "'") {
        const literal = readStringLiteral(source, cursor);
        return { nextIndex: scanToBoundary(source, literal.end), specifier: literal.value };
      }
      return { nextIndex: scanToBoundary(source, cursor), specifier: null };
    }

    const char = source[cursor];
    if (char === "\"" || char === "'" || char === "`") {
      cursor = readStringLiteral(source, cursor).end;
      continue;
    }

    if (char === ";" || char === "\n") {
      return { nextIndex: cursor + 1, specifier: null };
    }

    cursor += 1;
  }

  return { nextIndex: cursor, specifier: null };
}

function parseRequire(source: string, start: number): ParseResult {
  let cursor = skipWhitespaceAndComments(source, start + "require".length);
  if (source[cursor] !== "(") {
    return { nextIndex: cursor + 1, specifier: null };
  }

  cursor = skipWhitespaceAndComments(source, cursor + 1);
  if (source[cursor] === "\"" || source[cursor] === "'") {
    const literal = readStringLiteral(source, cursor);
    return { nextIndex: scanToBoundary(source, literal.end), specifier: literal.value };
  }

  return { nextIndex: scanToBoundary(source, cursor), specifier: null };
}

export function parseJsTsImports(source: string): string[] {
  const imports = new Set<string>();
  let index = 0;

  while (index < source.length) {
    index = skipWhitespaceAndComments(source, index);
    if (index >= source.length) {
      break;
    }

    const char = source[index];
    if (char === "\"" || char === "'" || char === "`") {
      index = readStringLiteral(source, index).end;
      continue;
    }

    if (startsWithToken(source, index, "import")) {
      const parsed = parseImport(source, index);
      if (parsed.specifier) imports.add(parsed.specifier);
      index = Math.max(parsed.nextIndex, index + 1);
      continue;
    }

    if (startsWithToken(source, index, "export")) {
      const parsed = parseExport(source, index);
      if (parsed.specifier) imports.add(parsed.specifier);
      index = Math.max(parsed.nextIndex, index + 1);
      continue;
    }

    if (startsWithToken(source, index, "require")) {
      const parsed = parseRequire(source, index);
      if (parsed.specifier) imports.add(parsed.specifier);
      index = Math.max(parsed.nextIndex, index + 1);
      continue;
    }

    index += 1;
  }

  return [...imports];
}

export function parseJavaImports(source: string): string[] {
  const imports = new Set<string>();
  const importPattern = /^\s*import\s+(?:static\s+)?([A-Za-z0-9_.*$]+)\s*;/gm;

  let match: RegExpExecArray | null;
  while ((match = importPattern.exec(source)) !== null) {
    const moduleName = match[1];
    if (moduleName) {
      imports.add(moduleName);
    }
  }

  return [...imports];
}

export function parsePythonImports(source: string): string[] {
  const imports = new Set<string>();
  const lines = source.split(/\r?\n/);

  for (const line of lines) {
    const importMatch = line.match(/^\s*import\s+([A-Za-z0-9_.,\s]+)\s*$/);
    if (importMatch) {
      const importBlock = importMatch[1];
      if (!importBlock) {
        continue;
      }

      for (const moduleName of importBlock.split(",")) {
        const trimmed = moduleName.trim();
        if (trimmed) imports.add(trimmed);
      }
      continue;
    }

    const fromMatch = line.match(/^\s*from\s+([A-Za-z0-9_.]+)\s+import\s+[A-Za-z0-9_.*,\s]+$/);
    if (fromMatch) {
      const fromPath = fromMatch[1];
      if (fromPath) {
        imports.add(fromPath);
      }
    }
  }

  return [...imports];
}

function cleanSpecifier(specifier: string): string | null {
  const trimmed = specifier.trim();
  if (!trimmed) {
    return null;
  }
  return trimmed;
}

export function parsePhpImports(source: string): string[] {
  const imports = new Set<string>();
  const includePattern =
    /\b(?:include|include_once|require|require_once)\s*(?:\(\s*)?["']([^"']+)["']\s*\)?/g;

  let match: RegExpExecArray | null;
  while ((match = includePattern.exec(source)) !== null) {
    const raw = match[1];
    if (!raw) {
      continue;
    }

    const specifier = cleanSpecifier(raw);
    if (specifier) {
      imports.add(specifier);
    }
  }

  return [...imports];
}

export function parseCssImports(source: string): string[] {
  const imports = new Set<string>();

  const cssImportPattern =
    /@import\s+(?:url\(\s*(?:["']([^"']+)["']|([^)\s]+))\s*\)|["']([^"']+)["'])/gi;

  let match: RegExpExecArray | null;
  while ((match = cssImportPattern.exec(source)) !== null) {
    const raw = match[1] ?? match[2] ?? match[3] ?? null;
    if (!raw) {
      continue;
    }

    const specifier = cleanSpecifier(raw);
    if (specifier) {
      imports.add(specifier);
    }
  }

  return [...imports];
}

export function parseHtmlImports(source: string): string[] {
  const imports = new Set<string>();
  const srcPattern = /<(?:script|img|source|iframe)\b[^>]*\bsrc\s*=\s*["']([^"']+)["'][^>]*>/gi;
  const hrefPattern = /<link\b[^>]*\bhref\s*=\s*["']([^"']+)["'][^>]*>/gi;

  const collect = (pattern: RegExp): void => {
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(source)) !== null) {
      const raw = match[1];
      if (!raw) {
        continue;
      }

      const specifier = cleanSpecifier(raw);
      if (!specifier) {
        continue;
      }

      imports.add(specifier);
    }
  };

  collect(srcPattern);
  collect(hrefPattern);

  return [...imports];
}

File: crates/ccv-analyzer/src/import_parser.rs
use regex::Regex;

use super::text::{read_string_literal, skip_whitespace_and_comments, starts_with_token};

struct ParseResult {
    next_index: usize,
    specifier: Option<String>,
}

fn scan_to_boundary(chars: &[char], start: usize) -> usize {
    let mut cursor = start;
    let mut depth = 0usize;

    while cursor < chars.len() {
        cursor = skip_whitespace_and_comments(chars, cursor);
        if cursor >= chars.len() {
            return cursor;
        }

        let current = chars[cursor];
        if current == '"' || current == '\'' || current == '`' {
            cursor = read_string_literal(chars, cursor).end;
            continue;
        }

        if current == '(' || current == '{' || current == '[' {
            depth += 1;
            cursor += 1;
            continue;
        }

        if current == ')' || current == '}' || current == ']' {
            depth = depth.saturating_sub(1);
            cursor += 1;
            continue;
        }

        if (current == ';' || current == '\n') && depth == 0 {
            return cursor + 1;
        }

        cursor += 1;
    }

    chars.len()
}

fn parse_import(chars: &[char], start: usize) -> ParseResult {
    let mut cursor = skip_whitespace_and_comments(chars, start + "import".len());

    if chars.get(cursor) == Some(&'(') {
        cursor = skip_whitespace_and_comments(chars, cursor + 1);
        if matches!(chars.get(cursor), Some('\'') | Some('"')) {
            let literal = read_string_literal(chars, cursor);
            return ParseResult {
                next_index: scan_to_boundary(chars, literal.end),
                specifier: Some(literal.value),
            };
        }

        return ParseResult {
            next_index: scan_to_boundary(chars, cursor),
            specifier: None,
        };
    }

    if matches!(chars.get(cursor), Some('\'') | Some('"')) {
        let literal = read_string_literal(chars, cursor);
        return ParseResult {
            next_index: scan_to_boundary(chars, literal.end),
            specifier: Some(literal.value),
        };
    }

    while cursor < chars.len() {
        cursor = skip_whitespace_and_comments(chars, cursor);
        if cursor >= chars.len() {
            break;
        }

        if starts_with_token(chars, cursor, "from") {
            cursor = skip_whitespace_and_comments(chars, cursor + "from".len());
            if matches!(chars.get(cursor), Some('\'') | Some('"')) {
                let literal = read_string_literal(chars, cursor);
                return ParseResult {
                    next_index: scan_to_boundary(chars, literal.end),
                    specifier: Some(literal.value),
                };
            }

            return ParseResult {
                next_index: scan_to_boundary(chars, cursor),
                specifier: None,
            };
        }

        let current = chars[cursor];
        if current == '"' || current == '\'' || current == '`' {
            cursor = read_string_literal(chars, cursor).end;
            continue;
        }

        if current == ';' || current == '\n' {
            return ParseResult {
                next_index: cursor + 1,
                specifier: None,
            };
        }

        cursor += 1;
    }

    ParseResult {
        next_index: cursor,
        specifier: None,
    }
}

fn parse_export(chars: &[char], start: usize) -> ParseResult {
    let mut cursor = skip_whitespace_and_comments(chars, start + "export".len());

    while cursor < chars.len() {
        cursor = skip_whitespace_and_comments(chars, cursor);
        if cursor >= chars.len() {
            break;
        }

        if starts_with_token(chars, cursor, "from") {
            cursor = skip_whitespace_and_comments(chars, cursor + "from".len());
            if matches!(chars.get(cursor), Some('\'') | Some('"')) {
                let literal = read_string_literal(chars, cursor);
                return ParseResult {
                    next_index: scan_to_boundary(chars, literal.end),
                    specifier: Some(literal.value),
                };
            }

            return ParseResult {
                next_index: scan_to_boundary(chars, cursor),
                specifier: None,
            };
        }

        let current = chars[cursor];
        if current == '"' || current == '\'' || current == '`' {
            cursor = read_string_literal(chars, cursor).end;
            continue;
        }

        if current == ';' || current == '\n' {
            return ParseResult {
                next_index: cursor + 1,
                specifier: None,
            };
        }

        cursor += 1;
    }

    ParseResult {
        next_index: cursor,
        specifier: None,
    }
}

fn parse_require(chars: &[char], start: usize) -> ParseResult {
    let mut cursor = skip_whitespace_and_comments(chars, start + "require".len());
    if chars.get(cursor) != Some(&'(') {
        return ParseResult {
            next_index: cursor.saturating_add(1),
            specifier: None,
        };
    }

    cursor = skip_whitespace_and_comments(chars, cursor + 1);
    if matches!(chars.get(cursor), Some('\'') | Some('"')) {
        let literal = read_string_literal(chars, cursor);
        return ParseResult {
            next_index: scan_to_boundary(chars, literal.end),
            specifier: Some(literal.value),
        };
    }

    ParseResult {
        next_index: scan_to_boundary(chars, cursor),
        specifier: None,
    }
}

fn clean_specifier(raw: &str) -> Option<String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return None;
    }
    Some(trimmed.to_string())
}

pub fn parse_js_ts_imports(source: &str) -> Vec<String> {
    let chars: Vec<char> = source.chars().collect();
    let mut imports = Vec::new();
    let mut seen = std::collections::BTreeSet::new();
    let mut index = 0usize;

    while index < chars.len() {
        index = skip_whitespace_and_comments(&chars, index);
        if index >= chars.len() {
            break;
        }

        let current = chars[index];
        if current == '"' || current == '\'' || current == '`' {
            index = read_string_literal(&chars, index).end;
            continue;
        }

        let parsed = if starts_with_token(&chars, index, "import") {
            Some(parse_import(&chars, index))
        } else if starts_with_token(&chars, index, "export") {
            Some(parse_export(&chars, index))
        } else if starts_with_token(&chars, index, "require") {
            Some(parse_require(&chars, index))
        } else {
            None
        };

        if let Some(result) = parsed {
            if let Some(specifier) = result.specifier {
                if seen.insert(specifier.clone()) {
                    imports.push(specifier);
                }
            }
            index = result.next_index.max(index + 1);
            continue;
        }

        index += 1;
    }

    imports
}

pub fn parse_java_imports(source: &str) -> Vec<String> {
    let pattern = Regex::new(r"(?m)^\s*import\s+(?:static\s+)?([A-Za-z0-9_.*$]+)\s*;")
        .expect("valid java import regex");
    pattern
        .captures_iter(source)
        .filter_map(|capture| capture.get(1).map(|value| value.as_str().to_string()))
        .collect::<std::collections::BTreeSet<_>>()
        .into_iter()
        .collect()
}

pub fn parse_python_imports(source: &str) -> Vec<String> {
    let mut imports = std::collections::BTreeSet::new();

    let import_pattern = Regex::new(r"^\s*import\s+([A-Za-z0-9_.,\s]+)\s*$")
        .expect("valid python import regex");
    let from_pattern = Regex::new(r"^\s*from\s+([A-Za-z0-9_.]+)\s+import\s+[A-Za-z0-9_.*,\s]+$")
        .expect("valid python from regex");

    for line in source.lines() {
        if let Some(capture) = import_pattern.captures(line) {
            if let Some(block) = capture.get(1) {
                for module_name in block.as_str().split(',') {
                    let trimmed = module_name.trim();
                    if !trimmed.is_empty() {
                        imports.insert(trimmed.to_string());
                    }
                }
            }
            continue;
        }

        if let Some(capture) = from_pattern.captures(line) {
            if let Some(module_name) = capture.get(1) {
                imports.insert(module_name.as_str().to_string());
            }
        }
    }

    imports.into_iter().collect()
}

pub fn parse_php_imports(source: &str) -> Vec<String> {
    let pattern = Regex::new(
        r#"\b(?:include|include_once|require|require_once)\s*(?:\(\s*)?["']([^"']+)["']\s*\)?"#,
    )
    .expect("valid php import regex");

    pattern
        .captures_iter(source)
        .filter_map(|capture| capture.get(1).and_then(|value| clean_specifier(value.as_str())))
        .collect::<std::collections::BTreeSet<_>>()
        .into_iter()
        .collect()
}

pub fn parse_css_imports(source: &str) -> Vec<String> {
    let pattern = Regex::new(
        r#"(?i)@import\s+(?:url\(\s*(?:["']([^"']+)["']|([^\)\s]+))\s*\)|["']([^"']+)["'])"#,
    )
    .expect("valid css import regex");

    pattern
        .captures_iter(source)
        .filter_map(|capture| {
            capture
                .get(1)
                .or_else(|| capture.get(2))
                .or_else(|| capture.get(3))
                .and_then(|value| clean_specifier(value.as_str()))
        })
        .collect::<std::collections::BTreeSet<_>>()
        .into_iter()
        .collect()
}

pub fn parse_html_imports(source: &str) -> Vec<String> {
    let src_pattern = Regex::new(
        r#"(?i)<(?:script|img|source|iframe)\b[^>]*\bsrc\s*=\s*["']([^"']+)["'][^>]*>"#,
    )
    .expect("valid html src regex");
    let href_pattern = Regex::new(r#"(?i)<link\b[^>]*\bhref\s*=\s*["']([^"']+)["'][^>]*>"#)
        .expect("valid html href regex");

    let mut imports = std::collections::BTreeSet::new();
    for pattern in [&src_pattern, &href_pattern] {
        for capture in pattern.captures_iter(source) {
            if let Some(specifier) = capture.get(1).and_then(|value| clean_specifier(value.as_str())) {
                imports.insert(specifier);
            }
        }
    }

    imports.into_iter().collect()
}

#[cfg(test)]
mod tests {
    use super::{parse_html_imports, parse_js_ts_imports};

    #[test]
    fn parses_js_ts_imports_and_ignores_comments_and_strings() {
        let source = r#"
import React from 'react';
// import nope from 'ignore-me';
const text = "require('fake')";
/* export * from 'hidden' */
const z = import("./dynamic");
const y = require('../legacy');
export { x } from "./utils";
import './styles.css';
"#;

        let mut parsed = parse_js_ts_imports(source);
        parsed.sort();

        assert_eq!(
            parsed,
            vec!["../legacy", "./dynamic", "./styles.css", "./utils", "react"]
        );
    }

    #[test]
    fn parses_html_imports() {
        let source = r#"
<html>
  <head>
    <link rel="stylesheet" href="./styles/main.css" />
    <script src="./scripts/app.js?v=1"></script>
  </head>
  <body>
    <img src="assets/logo.png" />
    <iframe src="https://example.com/embed"></iframe>
  </body>
</html>
"#;

        let parsed = parse_html_imports(source);
        assert_eq!(
            parsed,
            vec![
                "./scripts/app.js?v=1".to_string(),
                "./styles/main.css".to_string(),
                "assets/logo.png".to_string(),
                "https://example.com/embed".to_string(),
            ]
        );
    }
}
